utils/tail/tail.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tail implements "tail -F" functionality following rotated logs
+package tail
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/exp/inotify"
+)
+
+type Tail struct {
+	reader   *bufio.Reader
+	filename string
+	file     *os.File
+	stop     chan bool
+}
+
+const retryReopenInterval = time.Second
+const maxReopenAttempts = 3
+
+// NewTail starts opens the given file and watches it for deletion/rotation
+func NewTail(filename string) (t *Tail, err error) {
+	t = &Tail{
+		filename: filename,
+	}
+	t.file, err = os.Open(t.filename)
+	if err != nil {
+		glog.Errorf("Open failed on %s", t.filename)
+		return nil, err
+	}
+	// TODO: not interested in old events?
+	//t.file.Seek(0, os.SEEK_END)
+	t.reader = bufio.NewReader(t.file)
+	t.stop = make(chan bool)
+	go t.watchFile()
+	return t, nil
+}
+
+// Read implements the io.Reader interface for Tail
+func (t *Tail) Read(p []byte) (n int, err error) {
+	return t.reader.Read(p)
+}
+
+// type assertion
+var _ io.Reader = &Tail{}
+
+// Close stops watching and closes the file
+func (t *Tail) Close() {
+	t.file.Close()
+	close(t.stop)
+}
+
+func isEvent(event *inotify.Event, flag uint32) bool {
+	if event.Mask&flag == flag {
+		return true
+	}
+	return false
+}
+
+func (t *Tail) watchFile() {
+	watcher, err := inotify.NewWatcher()
+	if err != nil {
+		glog.Errorf("Inotify init failed: %s", err.Error())
+		return
+	}
+	for {
+		err := watcher.Watch(t.filename)
+		if err != nil {
+			glog.Errorf("Inotify watch failed: %s", err.Error())
+			return
```

Comment:
I think there's a lot that needs to be closed here, and similarly for returns below. I think this can all be refactored into several functions and rely on defers to close everything.